api/src/main/java/org/apache/iceberg/Accessors.java

```diff
@@ -187,6 +206,47 @@ public String toString() {
     }
   }
 
+  private static class DateAccessor extends PositionAccessor {
```

Comment:
These accessors are for types used by Iceberg generics. I don't think that we want to make them part of the public API, and we need to be able to add them when using generics but not use them when binding to other rows. These should not be used with `PartitionData` instances, for example.

Reply:
Looks like this is the cause of the test failures:

```
org.apache.iceberg.spark.source.TestFilteredScan > testDayPartitionedTimestampFilters[2] FAILED
    java.lang.IllegalArgumentException: Wrong class, java.time.LocalDate, for object: 17521
        at org.apache.iceberg.PartitionData.get(PartitionData.java:120)
        at org.apache.iceberg.Accessors$DateAccessor.get(Accessors.java:216)
        at org.apache.iceberg.Accessors$DateAccessor.get(Accessors.java:209)
        at org.apache.iceberg.expressions.BoundReference.eval(BoundReference.java:39)
```

We need to substitute this accessor for the default accessor only when using Iceberg generics.

Reply:
I agree that these accessors should be specific for Iceberg generics and basically there are three options to do that:

1. Identify Iceberg generic class in `Accessors` in iceberg-api, but that would mix in iceberg-data into iceberg-api package.

2. Copy all related `Accessors` into iceberg-data package, but `Evaluator` constructs accessors via `Schema#accessForField` which comes from iceberg-api package, so that we can not pass iceberg generic accessors to them. Or we need to update Evaluator but it is still in iceberg-api package.

3. Fix the issue: https://github.com/apache/incubator-iceberg/issues/984, so that DateAccessor works for PartitionData as well.

IMO, option 3 is feasible, what do you think? Or you may have other options.

Reply:
@rdblue , I just use another solution for this. Take `accessors` from passing parameter instead of parsing it from schema when binding expression with struct type. With this decoupling we could customize the `accessors` for different `StructLike`. Does that make sense to you?